htlcswitch/control_tower.go

```diff
@@ -90,67 +102,79 @@ func (p *paymentControl) ClearForTakeoff(htlc *lnwire.UpdateAddHTLC) error {
 			return err
 		}
 
-		// Reset the takeoff error, to avoid carrying over an error
-		// from a previous execution of the batched db transaction.
-		takeoffErr = nil
-
 		switch paymentStatus {
 
 		case channeldb.StatusGrounded:
 			// It is safe to reattempt a payment if we know that we
 			// haven't left one in flight. Since this one is
 			// grounded, Transition the payment status to InFlight
 			// to prevent others.
-			return channeldb.UpdatePaymentStatusTx(
+			err := channeldb.UpdatePaymentStatusTx(
 				tx, htlc.PaymentHash, channeldb.StatusInFlight,
 			)
+			if err != nil {
+				return err
+			}
+
+			err = channeldb.UpdatePaymentIDTx(
+				tx, htlc.PaymentHash, newPID,
+			)
+			if err != nil {
+				return err
+			}
+
+			pid = newPID
+			return nil
 
 		case channeldb.StatusInFlight:
-			// We already have an InFlight payment on the network. We will
-			// disallow any more payment until a response is received.
-			takeoffErr = ErrPaymentInFlight
+			// We already have an InFlight payment on the network.
+			// We will disallow any more payment until a response
+			// is received.
+			pid, err = channeldb.FetchPaymentIDTx(
+				tx, htlc.PaymentHash,
+			)
+			if err != nil {
+				return err
+			}
+			return ErrPaymentInFlight
 
 		case channeldb.StatusCompleted:
-			// We've already completed a payment to this payment hash,
-			// forbid the switch from sending another.
-			takeoffErr = ErrAlreadyPaid
+			// We've already completed a payment to this payment
+			// hash, forbid the switch from sending another.
+			preimage, err = channeldb.FetchPaymentPreimageTx(
```

Comment:
the side-effects of this transaction are not idempotent, which means we may see invalid reads from any of these subcommands. specifically, `pid` and `preimage` should be set to zero values each time the function is invoked

Reply:
Shouldn't matter, since `pid` is only used in case of `ErrPaymentInFlight` and `preimage` only in case of `ErrAlreadyPaid`. But probably doesn't hurt to do what you suggest.